pkg/api/validation/validation.go

```diff
@@ -1252,26 +1252,56 @@ func validateEndpointSubsets(subsets []api.EndpointSubset) errs.ValidationErrorL
 
 	for i := range subsets {
 		ss := &subsets[i]
-
-		ssErrs := errs.ValidationErrorList{}
-
 		if len(ss.Addresses) == 0 {
-			ssErrs = append(ssErrs, errs.NewFieldRequired("addresses"))
+			allErrs = append(allErrs, errs.NewFieldRequired("addresses"))
 		}
 		if len(ss.Ports) == 0 {
-			ssErrs = append(ssErrs, errs.NewFieldRequired("ports"))
+			allErrs = append(allErrs, errs.NewFieldRequired("ports"))
+		}
+		for addr := range ss.Addresses {
+			allErrs = append(allErrs, validateEndpointAddress(&ss.Addresses[addr])...)
 		}
-		// TODO: validate each address and port
-		allErrs = append(allErrs, ssErrs.PrefixIndex(i)...)
+		for port := range ss.Ports {
+			allErrs = append(allErrs, validateEndpointPort(&ss.Ports[port], len(ss.Ports) > 1)...)
+		}
+	}
+
+	return allErrs
+}
+
+func validateEndpointAddress(address *api.EndpointAddress) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	if !util.IsValidIPv4(address.IP) {
+		allErrs = append(allErrs, errs.NewFieldInvalid("IP", address.IP, "is invalid IPv4 address"))
 	}
+	return allErrs
+}
 
+func validateEndpointPort(port *api.EndpointPort, requireName bool) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	if requireName && port.Name == "" {
+		allErrs = append(allErrs, errs.NewFieldRequired("Name"))
+		return allErrs
```

Comment:
don't short-circuit - we want to report as many errors as we can in a single pass.